### PR TITLE
Apply global context for all apis

### PR DESCRIPTION
**Pull request by ebozduman**
This effort started with `mc ls <alias>` command issue, where user needed to hit CTRL-C twice to interrupt the command. 

The global context needed to be applied for `ls` and all other apis to have a complete solution.

There will be also a minio-go side fix that goes along with this effort. I'll also create that PR.

One manifestation of not using/having global context used by apis is the need for CTRL-C to be hit twice when the user wants the interrupt the api.

Example:
- Run `mc ls <alias>` when there is no MinIO server running. You CTRL-C once and the command will keep running until the second CTRL-C

### CODE REVIEW

**Comment by ebozduman** [comment]
#### Information about the  fix
The following is the list of commands, fix has touched. The ones without the `ctx context.Context` as their first argument are left alone, since global context is not used at all (`ShareUpload(bool, time.Duration, string)` and `GetURL()`). 
```
	Stat(ctx context.Context, isIncomplete, isPreserve bool, sse encrypt.ServerSide) (content *ClientContent, err *probe.Error)
	List(ctx context.Context, isRecursive, isIncomplete, isFetchMeta bool, showDir DirOpt) <-chan *ClientContent

	// Bucket operations
	MakeBucket(ctx context.Context, region string, ignoreExisting, withLock bool) *probe.Error
	// Object lock config
	SetObjectLockConfig(ctx context.Context, mode minio.RetentionMode, validity uint64, unit minio.ValidityUnit) *probe.Error
	GetObjectLockConfig(ctx context.Context) (mode minio.RetentionMode, validity uint64, unit minio.ValidityUnit, perr *probe.Error)

	// Access policy operations.
	GetAccess(ctx context.Context) (access string, policyJSON string, error *probe.Error)
	GetAccessRules(ctx context.Context) (policyRules map[string]string, error *probe.Error)
	SetAccess(ctx context.Context, access string, isJSON bool) *probe.Error

	// I/O operations
	Copy(ctx context.Context, source string, size int64, progress io.Reader, srcSSE, tgtSSE encrypt.ServerSide, metadata map[string]string, disableMultipart bool) *probe.Error

	// Runs select expression on object storage on specific files.
	Select(ctx context.Context, expression string, sse encrypt.ServerSide, opts SelectObjectOpts) (io.ReadCloser, *probe.Error)

	// I/O operations with metadata.
	Get(ctx context.Context, sse encrypt.ServerSide) (reader io.ReadCloser, err *probe.Error)
	Put(ctx context.Context, reader io.Reader, size int64, metadata map[string]string, progress io.Reader, sse encrypt.ServerSide, md5, disableMultipart bool) (n int64, err *probe.Error)

	// Object Locking related API
	PutObjectRetention(ctx context.Context, mode minio.RetentionMode, retainUntilDate time.Time, bypassGovernance bool) *probe.Error
	PutObjectLegalHold(ctx context.Context, hold minio.LegalHoldStatus) *probe.Error

	// I/O operations with expiration
	ShareDownload(ctx context.Context, expires time.Duration) (string, *probe.Error)
	ShareUpload(bool, time.Duration, string) (string, map[string]string, *probe.Error)

	// Watch events
	Watch(ctx context.Context, options WatchOptions) (*WatchObject, *probe.Error)

	// Delete operations
	Remove(ctx context.Context, isIncomplete, isRemoveBucket, isBypass bool, contentCh <-chan *ClientContent) (errorCh <-chan *probe.Error)

	AddUserAgent(ctx context.Context, app, version string)

	// Tagging operations
	GetTags(ctx context.Context) (*tags.Tags, *probe.Error)
	SetTags(ctx context.Context, tags string) *probe.Error
	DeleteTags(ctx context.Context) *probe.Error

	// Lifecycle operations
	GetLifecycle(ctx context.Context) (ilm.LifecycleConfiguration, *probe.Error)
	SetLifecycle(ctx context.Context, lfcCfg ilm.LifecycleConfiguration) *probe.Error
```

File: cmd/mirror-main.go [hunk 1]
```diff
@@ -776,27 +776,30 @@ func runMirror(srcURL, dstURL string, ctx *cli.Context, encKeyDB map[string][]pr
...
+		cli.Bool("md5"),
+		cli.Bool("disable-multipart"),
 	)
 
 	go func() {
 		<-globalContext.Done()
 		os.Exit(globalErrorExitStatus)
 	}()
 
+	ctx, cancelMirror := context.WithCancel(context.Background())
```
Comment (poornas) [inline cmd/mirror-main.go:800]:
```suggestion
	ctx, cancelMirror := context.WithCancel(globalContext)
```

File: cmd/watch-main.go [hunk 2]
```diff
@@ -157,8 +158,11 @@ func mainWatch(ctx *cli.Context) error {
 		Suffix:    suffix,
 	}
 
+	ctxt, cancelWatch := context.WithCancel(globalContext)
```
Comment (poornas) [inline cmd/watch-main.go:161]:
```suggestion
	ctx, cancelWatch := context.WithCancel(globalContext)
```

File: cmd/tree-main.go [hunk 3]
```diff
@@ -215,7 +216,7 @@ func doTree(url string, level int, leaf bool, branchString string, depth int, in
 		return nil
 	}
 
-	for content := range clnt.List(false, false, false, DirNone) {
+	for content := range clnt.List(globalContext, false, false, false, DirNone) {
```
Comment (poornas) [inline cmd/tree-main.go:219]:
```suggestion
	for content := range clnt.List(ctx, false, false, false, DirNone) {
```

File: cmd/sql-main.go [hunk 4]
```diff
@@ -464,13 +471,13 @@ func mainSQL(ctx *cli.Context) error {
 			continue
 		}
 
-		for content := range clnt.List(ctx.Bool("recursive"), false, false, DirNone) {
+		for content := range clnt.List(globalContext, cliCtx.Bool("recursive"), false, false, DirNone) {
```
Comment (poornas) [inline cmd/sql-main.go:474]:
```suggestion
		for content := range clnt.List(ctx, cliCtx.Bool("recursive"), false, false, DirNone) {
```

File: cmd/sql-main.go [hunk 5]
```diff
@@ -308,7 +309,7 @@ func getCSVHeader(sourceURL string, encKeyDB map[string][]prefixSSEPair) ([]stri
 	default:
 		var err *probe.Error
 		var metadata map[string]string
-		if r, metadata, err = getSourceStreamMetadataFromURL(sourceURL, encKeyDB); err != nil {
+		if r, metadata, err = getSourceStreamMetadataFromURL(context.Background(), sourceURL, encKeyDB); err != nil {
```
Comment (poornas) [inline cmd/sql-main.go:312]:
```suggestion
		if r, metadata, err = getSourceStreamMetadataFromURL(globalContext, sourceURL, encKeyDB); err != nil {
```

File: cmd/legalhold-main.go [hunk 6]
```diff
@@ -107,7 +111,7 @@ func setLegalHold(urlStr string, lhold minio.LegalHoldStatus, isRecursive bool)
 	alias, _, _ := mustExpandAlias(urlStr)
 	var cErr error
 	errorsFound := false
-	for content := range clnt.List(isRecursive, false, false, DirNone) {
+	for content := range clnt.List(globalContext, isRecursive, false, false, DirNone) {
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply cmd/legalhold-main.go]: Done

File: cmd/mirror-main.go (hunk 1 above)
Comment (ebozduman) [reply cmd/mirror-main.go]: Done

File: cmd/watch-main.go (hunk 2 above)
Comment (ebozduman) [reply cmd/watch-main.go]: Done

File: cmd/tree-main.go (hunk 3 above)
Comment (ebozduman) [reply cmd/tree-main.go]: Done

File: cmd/sql-main.go (hunk 4 above)
Comment (ebozduman) [reply cmd/sql-main.go]: Done

File: cmd/sql-main.go (hunk 5 above)
Comment (ebozduman) [reply cmd/sql-main.go]: Done

**Comment by BigUstad** [comment]
I'm not sure if this is an issue or expected behavior. I'd to press Ctrl-C twice for:
mc mb <unavailableAlias>/<bktname>
mc tag set <s3alias>/<existing_bkt>/<Non-existing_object> "key=value"
mc tag list <s3alias>/<existing_bkt>/<Non-existing_object>

**Comment by BigUstad** [comment]
> I'm not sure if this is an issue or expected behavior. I'd to press Ctrl-C twice for:
> mc mb /
> mc tag set /<existing_bkt>/<Non-existing_object> "key=value"
> mc tag list /<existing_bkt>/<Non-existing_object>

Working as expected. Please resolve conflict.

**Pull request merged**